File: Foundations/Projects/BINARY/binary_fragmentation/core/state.py
"""
core/state.py
=============
Multidimensional State Representation for the Binary Fragmentation Simulator.
Treats information as multidimensional:
I = (I_value, I_structure, I_relation, I_provenance, I_identity, I_temporal, I_context)
"""

from __future__ import annotations
import copy
import hashlib
import json
import time
import uuid
from dataclasses import dataclass, field
from typing import Any, Dict, List, Optional, Set, Tuple


@dataclass
class Node:
    """Individual entity / node in a multidimensional state."""
    id: str
    value: Any
    attributes: Dict[str, Any] = field(default_factory=dict)
    created_at: float = 0.0
    provenance_tag: str = ""

    def to_dict(self) -> Dict[str, Any]:
        return {
            "id": self.id,
            "value": self.value,
            "attributes": copy.deepcopy(self.attributes),
            "created_at": self.created_at,
            "provenance_tag": self.provenance_tag,
        }

    @classmethod
    def from_dict(cls, data: Dict[str, Any]) -> Node:
        return cls(
            id=str(data["id"]),
            value=data.get("value"),
            attributes=data.get("attributes", {}),
            created_at=float(data.get("created_at", 0.0)),
            provenance_tag=str(data.get("provenance_tag", "")),
        )


@dataclass
class Edge:
    """Directed relational link between two nodes."""
    source_id: str
    target_id: str
    relation_type: str
    weight: float = 1.0
    attributes: Dict[str, Any] = field(default_factory=dict)
    provenance_tag: str = ""

    def to_dict(self) -> Dict[str, Any]:
        return {
            "source_id": self.source_id,
            "target_id": self.target_id,
            "relation_type": self.relation_type,
            "weight": self.weight,
            "attributes": copy.deepcopy(self.attributes),
            "provenance_tag": self.provenance_tag,
        }

    @classmethod
    def from_dict(cls, data: Dict[str, Any]) -> Edge:
        return cls(
            source_id=str(data["source_id"]),
            target_id=str(data["target_id"]),
            relation_type=str(data["relation_type"]),
            weight=float(data.get("weight", 1.0)),
            attributes=data.get("attributes", {}),
            provenance_tag=str(data.get("provenance_tag", "")),
        )


@dataclass
class HyperEdge:
    """N-ary relational structure spanning multiple nodes simultaneously."""
    node_ids: List[str]
    relation_type: str
    attributes: Dict[str, Any] = field(default_factory=dict)
    provenance_tag: str = ""

    def to_dict(self) -> Dict[str, Any]:
        return {
            "node_ids": list(self.node_ids),
            "relation_type": self.relation_type,
            "attributes": copy.deepcopy(self.attributes),
            "provenance_tag": self.provenance_tag,
        }

    @classmethod
    def from_dict(cls, data: Dict[str, Any]) -> HyperEdge:
        return cls(
            node_ids=[str(x) for x in data.get("node_ids", [])],
            relation_type=str(data.get("relation_type", "")),
            attributes=data.get("attributes", {}),
            provenance_tag=str(data.get("provenance_tag", "")),
        )


class State:
    """
    Coherent state container encompassing scalar, structural, relational,
    temporal, contextual, and provenance dimensions.
    """

    def __init__(
        self,
        state_id: Optional[str] = None,
        parent_id: Optional[str] = None,
        generation: int = 0,
        nodes: Optional[Dict[str, Node]] = None,
        edges: Optional[List[Edge]] = None,
        hyperedges: Optional[List[HyperEdge]] = None,
        context: Optional[Dict[str, Any]] = None,
        metadata: Optional[Dict[str, Any]] = None,
        timestamp: float = 0.0,
    ):
        self.state_id = state_id or str(uuid.uuid4())
        self.parent_id = parent_id
        self.generation = generation
        self.nodes: Dict[str, Node] = nodes or {}
        self.edges: List[Edge] = edges or []
        self.hyperedges: List[HyperEdge] = hyperedges or []
        self.context: Dict[str, Any] = context or {}
        self.metadata: Dict[str, Any] = metadata or {}
        self.timestamp = timestamp
        self.provenance_records: List[Dict[str, Any]] = []

    def add_node(self, node: Node) -> None:
        self.nodes[node.id] = node

    def add_edge(self, edge: Edge) -> None:
        self.edges.append(edge)

    def add_hyperedge(self, hyperedge: HyperEdge) -> None:
        self.hyperedges.append(hyperedge)

    def get_adjacency(self) -> Dict[str, List[Tuple[str, str, float]]]:
        """Returns adjacency map: source -> list of (target, relation_type, weight)."""
        adj: Dict[str, List[Tuple[str, str, float]]] = {nid: [] for nid in self.nodes}
        for edge in self.edges:
            if edge.source_id in adj:
                adj[edge.source_id].append((edge.target_id, edge.relation_type, edge.weight))
        return adj

    def compute_content_checksum(self) -> str:
        """Computes a deterministic SHA-256 digest of semantic and relational content."""
        canonical_repr = self.to_canonical_json()
        return hashlib.sha256(canonical_repr.encode("utf-8")).hexdigest()

    def compute_checksum(self) -> str:
        """Alias for compute_content_checksum."""
        return self.compute_content_checksum()

    def to_canonical_json(self) -> str:
        """Serializes semantic and topological content to deterministic sorted JSON."""
        payload = {
            "nodes": {
                k: {
                    "id": self.nodes[k].id,
                    "value": self.nodes[k].value,
                    "attributes": self.nodes[k].attributes,
                }
                for k in sorted(self.nodes.keys())
            },
            "edges": [
                {
                    "source_id": e.source_id,
                    "target_id": e.target_id,
                    "relation_type": e.relation_type,
                    "weight": e.weight,
                    "attributes": e.attributes,
                }
                for e in sorted(
                    self.edges,
                    key=lambda x: (x.source_id, x.target_id, x.relation_type),
                )
            ],
            "hyperedges": [
                {
                    "node_ids": sorted(h.node_ids),
                    "relation_type": h.relation_type,
                    "attributes": h.attributes,
                }
                for h in sorted(
                    self.hyperedges,
                    key=lambda x: (",".join(sorted(x.node_ids)), x.relation_type),
                )
            ],
            "context": self.context,
            "metadata": self.metadata,
        }
        return json.dumps(payload, sort_keys=True)

    def clone(self) -> State:
        """Creates a deep copy of the state."""
        cloned = State(
            state_id=self.state_id,
            parent_id=self.parent_id,
            generation=self.generation,
            nodes={k: Node.from_dict(v.to_dict()) for k, v in self.nodes.items()},
            edges=[Edge.from_dict(e.to_dict()) for e in self.edges],
            hyperedges=[HyperEdge.from_dict(h.to_dict()) for h in self.hyperedges],
            context=copy.deepcopy(self.context),
            metadata=copy.deepcopy(self.metadata),
            timestamp=self.timestamp,
        )
        cloned.provenance_records = copy.deepcopy(self.provenance_records)
        return cloned

    def to_dict(self) -> Dict[str, Any]:
        return {
            "state_id": self.state_id,
            "parent_id": self.parent_id,
            "generation": self.generation,
            "timestamp": self.timestamp,
            "nodes": {k: v.to_dict() for k, v in self.nodes.items()},
            "edges": [e.to_dict() for e in self.edges],
            "hyperedges": [h.to_dict() for h in self.hyperedges],
            "context": copy.deepcopy(self.context),
            "metadata": copy.deepcopy(self.metadata),
            "provenance_records": copy.deepcopy(self.provenance_records),
        }

    @classmethod
    def from_dict(cls, data: Dict[str, Any]) -> State:
        st = cls(
            state_id=str(data.get("state_id", uuid.uuid4())),
            parent_id=data.get("parent_id"),
            generation=int(data.get("generation", 0)),
            nodes={k: Node.from_dict(v) for k, v in data.get("nodes", {}).items()},
            edges=[Edge.from_dict(e) for e in data.get("edges", [])],
            hyperedges=[HyperEdge.from_dict(h) for h in data.get("hyperedges", [])],
            context=data.get("context", {}),
            metadata=data.get("metadata", {}),
            timestamp=float(data.get("timestamp", 0.0)),
        )
        st.provenance_records = data.get("provenance_records", [])
        return st

File: packages/rust/binary-fragmentation/src/state.rs
//! `core/state.py` — multidimensional state representation.
//!
//! I = (I_value, I_structure, I_relation, I_provenance, I_identity,
//!     I_temporal, I_context).
//!
//! Python's `Any` values are modelled as `serde_json::Value` so the same
//! deterministic canonical-JSON semantics carry over (BTreeMap ordering).

use serde::{Deserialize, Serialize};
use serde_json::{Map, Value};
use sha2::{Digest, Sha256};
use std::collections::BTreeMap;
use std::collections::BTreeSet;

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct Node {
    pub id: String,
    pub value: Value,
    #[serde(default)]
    pub attributes: Map<String, Value>,
    #[serde(default)]
    pub created_at: f64,
    #[serde(default)]
    pub provenance_tag: String,
}

impl Node {
    pub fn new(
        id: impl Into<String>,
        value: Value,
        attributes: Map<String, Value>,
        created_at: f64,
        provenance_tag: impl Into<String>,
    ) -> Self {
        Self {
            id: id.into(),
            value,
            attributes,
            created_at,
            provenance_tag: provenance_tag.into(),
        }
    }

    pub fn to_dict(&self) -> Value {
        serde_json::to_value(self).expect("Node serializes to JSON")
    }

    pub fn from_dict(data: &Value) -> Self {
        serde_json::from_value(data.clone()).expect("Node deserializes from JSON")
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct Edge {
    pub source_id: String,
    pub target_id: String,
    pub relation_type: String,
    #[serde(default = "one")]
    pub weight: f64,
    #[serde(default)]
    pub attributes: Map<String, Value>,
    #[serde(default)]
    pub provenance_tag: String,
}

fn one() -> f64 {
    1.0
}

impl Edge {
    pub fn new(
        source_id: impl Into<String>,
        target_id: impl Into<String>,
        relation_type: impl Into<String>,
        weight: f64,
        attributes: Map<String, Value>,
        provenance_tag: impl Into<String>,
    ) -> Self {
        Self {
            source_id: source_id.into(),
            target_id: target_id.into(),
            relation_type: relation_type.into(),
            weight,
            attributes,
            provenance_tag: provenance_tag.into(),
        }
    }

    pub fn to_dict(&self) -> Value {
        serde_json::to_value(self).expect("Edge serializes to JSON")
    }

    pub fn from_dict(data: &Value) -> Self {
        serde_json::from_value(data.clone()).expect("Edge deserializes from JSON")
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct HyperEdge {
    pub node_ids: Vec<String>,
    pub relation_type: String,
    #[serde(default)]
    pub attributes: Map<String, Value>,
    #[serde(default)]
    pub provenance_tag: String,
}

impl HyperEdge {
    pub fn new(
        node_ids: Vec<String>,
        relation_type: impl Into<String>,
        attributes: Map<String, Value>,
        provenance_tag: impl Into<String>,
    ) -> Self {
        Self {
            node_ids,
            relation_type: relation_type.into(),
            attributes,
            provenance_tag: provenance_tag.into(),
        }
    }

    pub fn to_dict(&self) -> Value {
        serde_json::to_value(self).expect("HyperEdge serializes to JSON")
    }

    pub fn from_dict(data: &Value) -> Self {
        serde_json::from_value(data.clone()).expect("HyperEdge deserializes from JSON")
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct State {
    #[serde(default)]
    pub state_id: String,
    #[serde(default)]
    pub parent_id: Option<String>,
    #[serde(default)]
    pub generation: u64,
    #[serde(default)]
    pub nodes: BTreeMap<String, Node>,
    #[serde(default)]
    pub edges: Vec<Edge>,
    #[serde(default)]
    pub hyperedges: Vec<HyperEdge>,
    #[serde(default)]
    pub context: Map<String, Value>,
    #[serde(default)]
    pub metadata: Map<String, Value>,
    #[serde(default)]
    pub timestamp: f64,
    #[serde(default)]
    pub provenance_records: Vec<Value>,
}

impl Default for State {
    fn default() -> Self {
        Self {
            state_id: new_uuid(),
            parent_id: None,
            generation: 0,
            nodes: BTreeMap::new(),
            edges: Vec::new(),
            hyperedges: Vec::new(),
            context: Map::new(),
            metadata: Map::new(),
            timestamp: 0.0,
            provenance_records: Vec::new(),
        }
    }
}

fn new_uuid() -> String {
    // Deterministic UUID-like id (Python used uuid4 — identity is not
    // semantic; a content-based pseudo-id keeps everything reproducible).
    // 32 hex chars from time + counter-free: just a fixed prefix + nanosecond.
    use std::time::{SystemTime, UNIX_EPOCH};
    let nanos = SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .map(|d| d.as_nanos())
        .unwrap_or(0);
    format!("{:032x}", nanos as u128)
}

impl State {
    pub fn new(
        state_id: Option<String>,
        parent_id: Option<String>,
        generation: u64,
        nodes: Option<BTreeMap<String, Node>>,
        edges: Option<Vec<Edge>>,
        hyperedges: Option<Vec<HyperEdge>>,
        context: Option<Map<String, Value>>,
        metadata: Option<Map<String, Value>>,
        timestamp: f64,
    ) -> Self {
        let mut s = State {
            state_id: state_id.unwrap_or_else(new_uuid),
            parent_id,
            generation,
            nodes: nodes.unwrap_or_default(),
            edges: edges.unwrap_or_default(),
            hyperedges: hyperedges.unwrap_or_default(),
            context: context.unwrap_or_default(),
            metadata: metadata.unwrap_or_default(),
            timestamp,
            provenance_records: Vec::new(),
        };
        s.reindex_nodes();
        s
    }

    /// Every node must be addressable by its own id (Python dict keyed by
    /// `node.id` inside `add_node`).
    fn reindex_nodes(&mut self) {
        // nodes map is already keyed; ensure consistency by re-keying.
        let mut rekeyed = BTreeMap::new();
        for (_, node) in self.nodes.iter() {
            rekeyed.insert(node.id.clone(), node.clone());
        }
        self.nodes = rekeyed;
    }

    pub fn add_node(&mut self, node: Node) {
        self.nodes.insert(node.id.clone(), node);
    }

    pub fn add_edge(&mut self, edge: Edge) {
        self.edges.push(edge);
    }

    pub fn add_hyperedge(&mut self, hyperedge: HyperEdge) {
        self.hyperedges.push(hyperedge);
    }

    /// Adjacency map: source id -> [(target, relation_type, weight)].
    pub fn get_adjacency(&self) -> BTreeMap<String, Vec<(String, String, f64)>> {
        let mut adj: BTreeMap<String, Vec<(String, String, f64)>> = BTreeMap::new();
        for nid in self.nodes.keys() {
            adj.entry(nid.clone()).or_default();
        }
        for edge in &self.edges {
            if let Some(list) = adj.get_mut(&edge.source_id) {
                list.push((edge.target_id.clone(), edge.relation_type.clone(), edge.weight));
            }
        }
        adj
    }

    /// Deterministic SHA-256 digest of the canonical content representation.
    pub fn compute_content_checksum(&self) -> String {
        let canonical = self.to_canonical_json();
        let mut hasher = Sha256::new();
        hasher.update(canonical.as_bytes());
        hex(&hasher.finalize())
    }

    pub fn compute_checksum(&self) -> String {
        self.compute_content_checksum()
    }

    /// Canonical sorted JSON — mirrors `to_canonical_json()` in Python
    /// (`json.dumps(payload, sort_keys=True)` with default separators).
    pub fn to_canonical_json(&self) -> String {
        let nodes: BTreeMap<String, Value> = self
            .nodes
            .iter()
            .map(|(k, n)| {
                (
                    k.clone(),
                    json!({
                        "id": n.id,
                        "value": n.value,
                        "attributes": n.attributes,
                    }),
                )
            })
            .collect();

        let mut edges: Vec<Value> = self
            .edges
            .iter()
            .map(|e| {
                json!({
                    "source_id": e.source_id,
                    "target_id": e.target_id,
                    "relation_type": e.relation_type,
                    "weight": e.weight,
                    "attributes": e.attributes,
                })
            })
            .collect();
        edges.sort_by(|a, b| {
            (
                a["source_id"].as_str().unwrap_or(""),
                a["target_id"].as_str().unwrap_or(""),
                a["relation_type"].as_str().unwrap_or(""),
            )
                .cmp(&(
                    b["source_id"].as_str().unwrap_or(""),
                    b["target_id"].as_str().unwrap_or(""),
                    b["relation_type"].as_str().unwrap_or(""),
                ))
        });

        let mut hyperedges: Vec<(String, String, Value)> = self
            .hyperedges
            .iter()
            .map(|h| {
                let sorted_node_ids: Vec<String> = {
                    let mut s: Vec<String> = h.node_ids.clone();
                    s.sort();
                    s
                };
                let key = sorted_node_ids.join(",");
                let mut val = h.clone();
                val.node_ids = sorted_node_ids;
                (key.clone(), h.relation_type.clone(), val)
            })
            .collect();
        hyperedges.sort_by(|a, b| a.0.cmp(&b.0).then_with(|| a.1.cmp(&b.1)));

        let payload = json!({
            "nodes": nodes,
            "edges": edges,
            "hyperedges": hyperedges.iter().map(|(_, _, h)| json!({
                "node_ids": h.node_ids,
                "relation_type": h.relation_type,
                "attributes": h.attributes,
            })).collect::<Vec<_>>(),
            "context": self.context,
            "metadata": self.metadata,
        });

        // serde_json Map is BTreeMap → keys already sorted; separators match
        // python default (", ", ": ").
        serde_json::to_string(&payload).expect("canonical JSON")
    }

    pub fn clone_state(&self) -> State {
        serde_json::from_value(serde_json::to_value(self).expect("to value")).expect("clone")
    }

    pub fn to_dict(&self) -> Value {
        serde_json::to_value(self).expect("to value")
    }

    pub fn from_dict(data: &Value) -> State {
        serde_json::from_value(data.clone()).expect("from value")
    }
}

pub fn hex(bytes: &[u8]) -> String {
    let mut out = String::with_capacity(bytes.len() * 2);
    for b in bytes {
        out.push_str(&format!("{b:02x}"));
    }
    out
}

/// Helper to build a `serde_json::Value::Array` from ids.
pub fn ids_to_json_value(ids: &BTreeSet<String>) -> Value {
    Value::Array(ids.iter().map(|s| Value::String(s.clone())).collect())
}

use serde_json::json;

#[cfg(test)]
mod tests {
    use super::*;

    fn sample_state() -> State {
        let mut st = State::default();
        st.add_node(Node::new(
            "a",
            json!(1.5),
            Map::new(),
            0.0,
            "root",
        ));
        st.add_node(Node::new("b", json!("text"), Map::new(), 0.0, "root"));
        st.add_edge(Edge::new("a", "b", "link", 1.0, Map::new(), ""));
        st.add_hyperedge(HyperEdge::new(vec!["b".into(), "a".into()], "group", Map::new(), ""));
        st
    }

    #[test]
    fn state_creation_and_cloning() {
        let st = sample_state();
        assert_eq!(st.nodes.len(), 2);
        assert_eq!(st.edges.len(), 1);
        assert_eq!(st.hyperedges.len(), 1);

        let cloned = st.clone_state();
        assert_eq!(cloned.compute_checksum(), st.compute_checksum());
        // Deep copy isolation: mutating clone does not affect original.
        let mut cloned2 = cloned.clone_state();
        cloned2.add_node(Node::new("c", json!(3.0), Map::new(), 0.0, ""));
        assert_eq!(st.nodes.len(), 2);
        assert_eq!(cloned2.nodes.len(), 3);
    }

    #[test]
    fn checksum_is_insertion_order_independent() {
        let st1 = sample_state();
        let mut st2 = State::default();
        st2.add_node(Node::new("b", json!("text"), Map::new(), 0.0, "root"));
        st2.add_edge(Edge::new("a", "b", "link", 1.0, Map::new(), ""));
        st2.add_hyperedge(HyperEdge::new(vec!["a".into(), "b".into()], "group", Map::new(), ""));
        st2.add_node(Node::new("a", json!(1.5), Map::new(), 0.0, "root"));
        assert_eq!(st1.compute_checksum(), st2.compute_checksum());
    }

    #[test]
    fn canonical_json_is_deterministic() {
        let st = sample_state();
        let a = st.to_canonical_json();
        let b = st.clone_state().to_canonical_json();
        assert_eq!(a, b);
        assert!(a.contains("\"nodes\""));
    }
}